crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,542 @@
+            )
+            .unwrap()
+            .0
+            .into_inner();
+
+            let value_clone = values.clone();
+
+            let mut param = [Pod::from_bytes(&values).unwrap()];
+
+            let props = properties! {
+                *pw::keys::MEDIA_TYPE => "Audio",
+                *pw::keys::MEDIA_CATEGORY => "Playback",
+            };
+
+            let stream = pw::stream::Stream::new(&self.core, "audio-output", props)
+                .expect("could not create new stream");
+
+            let streams = self.stream_params.clone();
+
+            let listener_stream = stream
+                .add_local_listener()
+                .state_changed(move |old, new| {
+                    debug!("State changed: {:?} -> {:?}", old, new);
+                })
+                .param_changed(move |stream, id, _data, param| {
+                    let Some(_param) = param else {
+                        return;
+                    };
+                    if id != ParamType::Format.as_raw() {
+                        return;
+                    }
+                    let mut param = [Pod::from_bytes(&value_clone).unwrap()];
+
+                    //callback to negotiate new set of streams
+                    stream
+                        .update_params(&mut param)
+                        .expect("could not update params");
+                })
+                .process(move |stream, _data| {
+                    //todo: use safe dequeue_buffer(), contribute queue_buffer()
+                    unsafe {
+                        let b: *mut pw_buffer = stream.dequeue_raw_buffer();
+                        if b.is_null() {
+                            return;
+                        }
+                        let buf = (*b).buffer;
+                        let datas = (*buf).datas;
+
+                        let p = (*datas).data as *mut u8;
+
+                        if p.is_null() {
+                            return;
+                        }
+
+                        // to calculate as sizeof(int16_t) * NR_CHANNELS
+                        let frame_size = info.channels as u32 * size_of::<i16>() as u32;
+                        let req = (*b).requested * (frame_size as u64);
+                        let mut n_bytes = cmp::min(req as u32, (*datas).maxsize);
+
+                        let mut streams = streams.write().unwrap();
+                        let streams = streams
+                            .get_mut(stream_id as usize)
+                            .expect("Stream does not exist");
+                        let Some(buffer) = streams.buffers.front_mut() else {
+                            return;
+                        };
+
+                        let mut start = buffer.pos;
+
+                        let avail = (buffer.bytes.len() - start) as i32;
+
+                        if avail <= 0 {
+                            // pad with silence
+                            ptr::write_bytes(p, 0, n_bytes as usize);
```

Comment:
is this the correct function to build silent?

Reply:
in pipewire C API they use memset for silence, so i found the equivalent in rust to be this.